

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>src.evo package &mdash; FEAGI 0.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="src.inf package" href="src.inf.html" />
    <link rel="prev" title="src.edu package" href="src.edu.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> FEAGI
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">FEAGI</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="setup.html">setup module</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="src.html">src package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="src.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="src.edu.html">src.edu package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">src.evo package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.inf.html">src.inf package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.ipu.html">src.ipu package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.mem.html">src.mem package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.npu.html">src.npu package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.opu.html">src.opu package</a></li>
<li class="toctree-l4"><a class="reference internal" href="src.vis.html">src.vis package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="src.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="src.html#module-src.json_to_ini">src.json_to_ini module</a></li>
<li class="toctree-l3"><a class="reference internal" href="src.html#module-src.main">src.main module</a></li>
<li class="toctree-l3"><a class="reference internal" href="src.html#src-skeleton-module">src.skeleton module</a></li>
<li class="toctree-l3"><a class="reference internal" href="src.html#module-src">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">FEAGI</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="modules.html">FEAGI</a> &raquo;</li>
        
          <li><a href="src.html">src package</a> &raquo;</li>
        
      <li>src.evo package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/src.evo.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="src-evo-package">
<h1>src.evo package<a class="headerlink" href="#src-evo-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-src.evo.death">
<span id="src-evo-death-module"></span><h2>src.evo.death module<a class="headerlink" href="#module-src.evo.death" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt id="src.evo.death.register">
<code class="sig-prename descclassname">src.evo.death.</code><code class="sig-name descname">register</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.death.register" title="Permalink to this definition">¶</a></dt>
<dd><p>Document test</p>
<dl class="simple">
<dt>Args:</dt><dd><p>ABCD</p>
</dd>
<dt>Returns:</dt><dd><p>Something</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-src.evo.genetics">
<span id="src-evo-genetics-module"></span><h2>src.evo.genetics module<a class="headerlink" href="#module-src.evo.genetics" title="Permalink to this headline">¶</a></h2>
<p>Copyright (c) 2019 Mohammad Nadji-Tehrani &lt;<a class="reference external" href="mailto:m&#46;nadji&#46;tehrani&#37;&#52;&#48;gmail&#46;com">m<span>&#46;</span>nadji<span>&#46;</span>tehrani<span>&#64;</span>gmail<span>&#46;</span>com</a>&gt;</p>
<p>This module contains functions related to genome handling mainly inspired by nature and genetic algorithms.</p>
<p>Evolution
- Generational</p>
<blockquote>
<div><ul class="simple">
<li><p>Gene mutation</p></li>
<li><p>Crossover</p></li>
<li><p>Genome selection</p></li>
<li><p>Pairing</p></li>
<li><p>Colonies, groups, and boundaries</p></li>
<li><p>Epigenetics and selective enablement of genes</p></li>
<li><p>Passing of gene changes to next generation</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>Developmental</dt><dd><ul>
<li><dl class="simple">
<dt>Global</dt><dd><ul>
<li><p>System level optimization</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Localized</dt><dd><ul>
<li><p>Within a cortical later</p></li>
<li><p>Within a cortical pathway</p></li>
<li><p>Within an organ</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Reflection of stress on a region back on genome</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<p>todo: Gene regulatory considerations</p>
<dl class="py class">
<dt id="src.evo.genetics.GeneModifier">
<em class="property">class </em><code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">GeneModifier</code><a class="headerlink" href="#src.evo.genetics.GeneModifier" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_consecutive_fire_cnt_max">
<em class="property">static </em><code class="sig-name descname">change_consecutive_fire_cnt_max</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_consecutive_fire_cnt_max" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron consecutive_fire_cnt_max in a given cortical area</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_cortical_dimensions">
<em class="property">static </em><code class="sig-name descname">change_cortical_dimensions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_cortical_dimensions" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the size of a cortical area’s dimension</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_cortical_neuron_count">
<em class="property">static </em><code class="sig-name descname">change_cortical_neuron_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_cortical_neuron_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron density aka Cortical Neuron Count in a given cortical area</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_depolarization_timer_threshold">
<em class="property">static </em><code class="sig-name descname">change_depolarization_timer_threshold</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_depolarization_timer_threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron timer threshold in a given cortical area</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_firing_threshold">
<em class="property">static </em><code class="sig-name descname">change_firing_threshold</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_firing_threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron firing threshold in a given cortical area</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_growth_rule_4_param_2">
<em class="property">static </em><code class="sig-name descname">change_growth_rule_4_param_2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_growth_rule_4_param_2" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron snooze_length in a given cortical area</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_snooze_length">
<em class="property">static </em><code class="sig-name descname">change_snooze_length</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_snooze_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron snooze_length in a given cortical area</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier.change_vision_plasticity_constant">
<em class="property">static </em><code class="sig-name descname">change_vision_plasticity_constant</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">change_percentage</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier.change_vision_plasticity_constant" title="Permalink to this definition">¶</a></dt>
<dd><p>Function to increase or decrease the neuron snooze_length in a given cortical area</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="src.evo.genetics.GeneModifier3">
<em class="property">class </em><code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">GeneModifier3</code><a class="headerlink" href="#src.evo.genetics.GeneModifier3" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt id="src.evo.genetics.GeneModifier3.mutate">
<code class="sig-name descname">mutate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier3.mutate" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="src.evo.genetics.GeneModifier3.random_gene_selector">
<code class="sig-name descname">random_gene_selector</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome_segment</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.GeneModifier3.random_gene_selector" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.calculate_brain_cognitive_fitness">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">calculate_brain_cognitive_fitness</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">test_stats</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.calculate_brain_cognitive_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the effectiveness of a given genome:
1. Fitness value will be a number between 0 and 100 with 100 the highest fitness possible (how can there be limit?)
2. Brain activeness should be considered as a factor. Brain that only guessed one number and that one correctly</p>
<blockquote>
<div><p>should not be considered better than a brain that guessed 100s of numbers 95% correct.</p>
</div></blockquote>
<ol class="arabic simple" start="3">
<li><p>Number of guess attempts vs. number of correct guesses is a factor</p></li>
</ol>
<p>Fitness calculation formula:</p>
<p>TE = Total number of times brain has been exposed to a character
TC = Total number of times brain has comprehended a number correctly
AF = Activity factor that is measured by a threshold.</p>
<blockquote>
<div><ul class="simple">
<li><p>&lt; 10 exposures leads to 0</p></li>
<li><p>&gt; 10 &amp; &lt; 50 exposures leads to n / 50 factor</p></li>
<li><p>&gt; 50 exposures leads to 1</p></li>
</ul>
</div></blockquote>
<p>TC / TE = Percentage of correct comprehensions</p>
<p>Genome fitness factor = AF * TC / TE</p>
<p>todo: Investigate cases where looking at one number stimulates multiple numbers together</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.calculate_survival_prob">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">calculate_survival_prob</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.calculate_survival_prob" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.compare_genomes">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">compare_genomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.compare_genomes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.crossover">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">crossover</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.crossover" title="Permalink to this definition">¶</a></dt>
<dd><p>To corssover genome 1 and 2, first list of keys from one genome is read and the content of that key
is swapped with the other genome.</p>
<p>todo: Given genome is hierarchical, crossover need to account for different levels</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.gene_anomaly_detector">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">gene_anomaly_detector</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.gene_anomaly_detector" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.generation_assessment">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">generation_assessment</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.generation_assessment" title="Permalink to this definition">¶</a></dt>
<dd><p>A collection of assessments to evaluate the performance of the Genome</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.genethesize">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">genethesize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.genethesize" title="Permalink to this definition">¶</a></dt>
<dd><p>Responsible for generating a complete set of Genome</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.genome_id_gen">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">genome_id_gen</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">size</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">chars</span><span class="o">=</span><span class="default_value">'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.genome_id_gen" title="Permalink to this definition">¶</a></dt>
<dd><p>This function generates a unique id which will be associated with each GEnome</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.genome_stats_analytics">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">genome_stats_analytics</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">test_stats</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.genome_stats_analytics" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.get_genome_candidate">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">get_genome_candidate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.get_genome_candidate" title="Permalink to this definition">¶</a></dt>
<dd><p>Scans genome db for a high performing genome and returns one from top 10% by random</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.highest_fitness_genome">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">highest_fitness_genome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.highest_fitness_genome" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.latest_genome">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">latest_genome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.latest_genome" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.mutate">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">mutate</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.mutate" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.random_genome">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">random_genome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.random_genome" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.select_a_genome">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">select_a_genome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.select_a_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>This function randomly selects a genetic operation from following options to produce a genome:
1. Crossover two randomly selected genomes
2. Random selection from available genomes
3. Latest genome
4. Best fitness ever
5. Mutate genome with highest fitness
6. TBD</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.selection">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">selection</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.selection" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.spin_new_generation">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">spin_new_generation</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.spin_new_generation" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.synthesize_new_gen">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">synthesize_new_gen</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.synthesize_new_gen" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.genetics.translate_genotype2phenotype">
<code class="sig-prename descclassname">src.evo.genetics.</code><code class="sig-name descname">translate_genotype2phenotype</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.genetics.translate_genotype2phenotype" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-src.evo.neuroembryogenesis">
<span id="src-evo-neuroembryogenesis-module"></span><h2>src.evo.neuroembryogenesis module<a class="headerlink" href="#module-src.evo.neuroembryogenesis" title="Permalink to this headline">¶</a></h2>
<p>Copyright (c) 2019 Mohammad Nadji-Tehrani &lt;<a class="reference external" href="mailto:m&#46;nadji&#46;tehrani&#37;&#52;&#48;gmail&#46;com">m<span>&#46;</span>nadji<span>&#46;</span>tehrani<span>&#64;</span>gmail<span>&#46;</span>com</a>&gt;</p>
<p>This module is responsible for reading instructions from genome as a genotype and translating them into connectome as
a phenotype.</p>
<p>Instructions in this module are biologically inspired and a high level model after the neuroembryogenesis process that
translates genome to a fully developed brain starting from the embryo’s neural tube.</p>
<p>In this framework, genome contains a list of all cortical layers and sub-layers for the brain. Each layer or sub-layer
has properties embedded within genome. Such properties are translated using this module to perform neurogenesis that
creates neurons followed by the process of synaptogenesis that is responsible for the creation of the connectivity
among neurons. All of the structures generated as a result of this process is stored in a data structure called
connectome.</p>
<dl class="py function">
<dt id="src.evo.neuroembryogenesis.build_cortical_map">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">build_cortical_map</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.build_cortical_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Develops a data structure suitable for graphing cortical connectivity using iGraph package</p>
<p>Sample output structure
graph_edges = [(0, 2), (0, 1), (0, 3), (1, 2), (1, 3), (2, 4), (3, 4), (3, 0)]
graph_weights = [8, 6, 3, 5, 6, 4, 9, 50]
graph_labels = [‘v1’, ‘v2’, ‘v3’, ‘v4’, ‘v5’]</p>
<p>Returns: Cortical map in a … format.</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.build_synapse_intercortical">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">build_synapses</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">brain</span></em>, <em class="sig-param"><span class="n">parameters</span></em>, <em class="sig-param"><span class="n">block_dic</span></em>, <em class="sig-param"><span class="n">src_cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.build_synapse_intercortical" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.build_synapse_intracortical">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">build_synapse_intracortical</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">brain</span></em>, <em class="sig-param"><span class="n">parameters</span></em>, <em class="sig-param"><span class="n">key</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.build_synapse_intracortical" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.connectome_backup">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">connectome_backup</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">src</span></em>, <em class="sig-param"><span class="n">dst</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.connectome_backup" title="Permalink to this definition">¶</a></dt>
<dd><p>Backs up an existing connectome to preserve all the structures and associated memory data for future use.
Args:</p>
<blockquote>
<div><p>src (String): Location of the source connectome folder
dst (String): Destination folder for storing connectome backup</p>
</div></blockquote>
<p>Returns:</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.connectome_structural_fitness">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">connectome_structural_fitness</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.connectome_structural_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>To conduct a set of validations and calculate a structural fitness for the developed connectome. The returned value
can be used to eliminate a premature structure.</p>
<p>Returns: Structural fitness factor</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.cortical_connectivity_eval">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">cortical_connectivity_eval</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.cortical_connectivity_eval" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluates whether the cortical structure is meeting minimum expectations for basic functionality.</p>
<p>Returns: True or False</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.cortical_list">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">cortical_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.cortical_list" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.cortical_sub_group_members">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">cortical_sub_group_members</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">group</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.cortical_sub_group_members" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.develop">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">develop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.develop" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.develop_brain">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">develop_brain</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reincarnation_mode</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.develop_brain" title="Permalink to this definition">¶</a></dt>
<dd><p>This function operates in two modes. If run with reincarnation mode as false then it will develop a new brain by
reading instructions from genome and developing it step by step. If run with reincarnation mode as True then the
last available connectome data will be used to operate as a new brain. Reincarnation set to True will preserve the
memories and learning from the previous generation while starting a new brain instance.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>reincarnation_mode (bool): If true, an existing connectome will be used to reuse an existing brain structure.</p>
</dd>
</dl>
<p>Returns:</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.ipu_opu_connectivity">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">ipu_opu_connectivity</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.ipu_opu_connectivity" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluates synaptic connectivity between IPU and OPU</p>
<p>Returns: True or False</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.neurogenesis">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">neurogenesis</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.neurogenesis" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.neuron_count">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">neuron_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.neuron_count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.reset_connectome_files">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">reset_connectome_files</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.reset_connectome_files" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.reset_connectome_in_mem">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">reset_connectome_in_mem</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.reset_connectome_in_mem" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.reuse">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">reuse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.reuse" title="Permalink to this definition">¶</a></dt>
<dd><p>Placeholder for a function to reuse an existing connectome.</p>
<p>Returns:</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.synapse_count">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">synapse_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area_src</span></em>, <em class="sig-param"><span class="n">cortical_area_dst</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.synapse_count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuroembryogenesis.synaptogenesis">
<code class="sig-prename descclassname">src.evo.neuroembryogenesis.</code><code class="sig-name descname">synaptogenesis</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuroembryogenesis.synaptogenesis" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-src.evo.stats">
<span id="src-evo-stats-module"></span><h2>src.evo.stats module<a class="headerlink" href="#module-src.evo.stats" title="Permalink to this headline">¶</a></h2>
<p>Provides functions performing statistical analysis on the Connectome and Cortical behavior</p>
<dl class="py function">
<dt id="src.evo.stats.brain_total_synapse_cnt">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">brain_total_synapse_cnt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">verbose</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.brain_total_synapse_cnt" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.candidate_list_counter">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">candidate_list_counter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">candidate_list</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.candidate_list_counter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.connectome_neighbor_histogram">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">connectome_neighbor_histogram</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.connectome_neighbor_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots a Histogram of count of neighbor relationships per Neuron in a Cortical area</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.stats.connectome_neuron_count">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">connectome_neuron_count</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.connectome_neuron_count" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.connectome_total_synapse_cnt">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">connectome_total_synapse_cnt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.connectome_total_synapse_cnt" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total number of Neurons and Synapses for a given cortical area</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.stats.cortical_area_neuron_count">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">cortical_area_neuron_count</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.cortical_area_neuron_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of Neurons in the connectome</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.stats.cortical_xyz_range">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">cortical_xyz_range</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.cortical_xyz_range" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.fcl_stats">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">fcl_stats</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome_id</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.fcl_stats" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.list_common_upstream_neurons">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">list_common_upstream_neurons</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">neuron_a</span></em>, <em class="sig-param"><span class="n">neuron_b</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.list_common_upstream_neurons" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.list_top_n_utf_memory_neurons">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">list_top_n_utf_memory_neurons</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">n</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.list_top_n_utf_memory_neurons" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.list_upstream_neuron_count_for_digits">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">list_upstream_neuron_count_for_digits</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">digit</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.list_upstream_neuron_count_for_digits" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.print_cortical_stats">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">print_cortical_stats</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.print_cortical_stats" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.stats.print_fcl_stats">
<code class="sig-prename descclassname">src.evo.stats.</code><code class="sig-name descname">print_fcl_stats</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">genome_id</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.stats.print_fcl_stats" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-src.evo.neuron">
<span id="src-evo-neuron-module"></span><h2>src.evo.neuron module<a class="headerlink" href="#module-src.evo.neuron" title="Permalink to this headline">¶</a></h2>
<p>A collection of functions related to Neurons</p>
<dl class="py function">
<dt id="src.evo.neuron.block_id_gen">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">block_id_gen</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">coordinate</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.block_id_gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Generating a block id so it can be used for faster neighbor detection</p>
<p>Args:</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>Something</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.block_reference_builder">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">block_reference_builder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">block</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.block_reference_builder" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_apoptosis">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_apoptosis</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_apoptosis" title="Permalink to this definition">¶</a></dt>
<dd><p>Responsible for programmed death of neuron</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_create">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_create</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">soma_location</span></em>, <em class="sig-param"><span class="n">dendrite_locations</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_create" title="Permalink to this definition">¶</a></dt>
<dd><p>Responsible for adding a Neuron to connectome</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_dendrite_location_generator">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_dendrite_location_generator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">neuron_location</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_dendrite_location_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>generates location and block information of the neuron dendrites</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_dendrite_template">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_dendrite_template</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">direction</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_dendrite_template" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_finder">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_finder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">location</span></em>, <em class="sig-param"><span class="n">radius</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_finder" title="Permalink to this definition">¶</a></dt>
<dd><p>Queries a given cortical area and returns a listed of Neuron IDs matching search criteria</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_genesis_3d">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_genesis_3d</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_genesis_3d" title="Permalink to this definition">¶</a></dt>
<dd><p>Function responsible for creating Neurons using the blueprint</p>
<ol class="arabic simple">
<li><p>location_collector function generates the list of all neuron locations for a given cortical area</p></li>
<li><p>for each location, create the list of dendrite locations and associated block</p></li>
</ol>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_id_gen">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_id_gen</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">size</span><span class="o">=</span><span class="default_value">6</span></em>, <em class="sig-param"><span class="n">chars</span><span class="o">=</span><span class="default_value">'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_id_gen" title="Permalink to this definition">¶</a></dt>
<dd><p>This function generates a unique id which will be associated with each neuron
:param size:
:param chars:
:return:</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_location_collector">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_location_collector</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_location_collector" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a list of locations to be used for Neuron creation
:return:</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.neuron.neuron_location_gen">
<code class="sig-prename descclassname">src.evo.neuron.</code><code class="sig-name descname">neuron_location_gen</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">x1</span></em>, <em class="sig-param"><span class="n">y1</span></em>, <em class="sig-param"><span class="n">z1</span></em>, <em class="sig-param"><span class="n">x2</span></em>, <em class="sig-param"><span class="n">y2</span></em>, <em class="sig-param"><span class="n">z2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.neuron.neuron_location_gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Function responsible to generate a pseudo-random location for a Neuron given some constraints</p>
</dd></dl>

</div>
<div class="section" id="module-src.evo.synapse">
<span id="src-evo-synapse-module"></span><h2>src.evo.synapse module<a class="headerlink" href="#module-src.evo.synapse" title="Permalink to this headline">¶</a></h2>
<p>This module covers needed tools for synapse creation.</p>
<dl class="py class">
<dt id="src.evo.synapse.SynaptogenesisRuleManager">
<em class="property">class </em><code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">SynaptogenesisRuleManager</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">src_neuron_id</span></em>, <em class="sig-param"><span class="n">src_cortical_area</span></em>, <em class="sig-param"><span class="n">dst_cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.SynaptogenesisRuleManager" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Manages the rules involved with neuron synaptogenesis process.</p>
<p>Needed info: Rule and rule params
return: the func for the rule</p>
<dl class="py method">
<dt id="src.evo.synapse.SynaptogenesisRuleManager.growth_rule_selector">
<code class="sig-name descname">growth_rule_selector</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.SynaptogenesisRuleManager.growth_rule_selector" title="Permalink to this definition">¶</a></dt>
<dd><p>Provides a mapping table between rules ids defined in genome and the functions driving the rules
Args:</p>
<blockquote>
<div><p>rule_id: as referenced within Genome</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>Function corresponding the rule_id</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="src.evo.synapse.SynaptogenesisRuleManager.rule_block_to_block">
<code class="sig-name descname">rule_block_to_block</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.SynaptogenesisRuleManager.rule_block_to_block" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the list of neurons listed under the same block id as the source neuron
todo: need to add ability to account for the sublayer mapping e.g. v1.1 &gt; v2 layer 1 based on z offset
todo: instead of a single destination block based on the rule pass a set of blocks to make it generalizable
Returns:</p>
</dd></dl>

<dl class="py method">
<dt id="src.evo.synapse.SynaptogenesisRuleManager.rule_neuron_to_neuron">
<code class="sig-name descname">rule_neuron_to_neuron</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.SynaptogenesisRuleManager.rule_neuron_to_neuron" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.block_ref_2_id">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">block_ref_2_id</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">block_ref</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.block_ref_2_id" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.block_z_offset">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">block_z_offset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">block_ref</span></em>, <em class="sig-param"><span class="n">offset</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.block_z_offset" title="Permalink to this definition">¶</a></dt>
<dd><p>Offsets the z coordinate of the block reference by the value defined by “offset”
todo: There is a risk that the new offset value exceed defined block boundaries of a given cortical area</p>
<dl class="simple">
<dt>Args:</dt><dd><p>block_ref:
offset:</p>
</dd>
</dl>
<p>Returns: Adjusted block reference</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.cortical_area_lengths">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">cortical_area_lengths</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.cortical_area_lengths" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neighbor_builder_intercortical">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neighbor_builder_intercortical</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">brain</span></em>, <em class="sig-param"><span class="n">genome</span></em>, <em class="sig-param"><span class="n">brain_gen</span></em>, <em class="sig-param"><span class="n">cortical_area_dst</span></em>, <em class="sig-param"><span class="n">rule</span></em>, <em class="sig-param"><span class="n">rule_param</span></em>, <em class="sig-param"><span class="n">postsynaptic_current</span><span class="o">=</span><span class="default_value">1.1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neighbor_builder_intercortical" title="Permalink to this definition">¶</a></dt>
<dd><p>Crawls thru a Cortical area/layer and builds Synapses with another Cortical area/layer</p>
<p>todo: take advantage of multi processing building the synapses for a given cortical area
todo: deficiency when brain gen is false</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neighbor_candidate_generator">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neighbor_candidate_generator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">src_cortical_area</span></em>, <em class="sig-param"><span class="n">src_neuron_id</span></em>, <em class="sig-param"><span class="n">dst_cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neighbor_candidate_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Identifies the list of candidate neurons in the destination cortical area that based on the rules defined by
source cortical area are suitable fit for synapse creation.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>src_cortical_area:
src_neuron_id:
dst_cortical_area:</p>
</dd>
<dt>Returns:</dt><dd><p>List of candidate Neurons</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neighbor_finder_intercortical">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neighbor_finder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">cortical_area_dst</span></em>, <em class="sig-param"><span class="n">src_neuron_id</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neighbor_finder_intercortical" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds a list of candidate Neurons from another Cortical area to build Synapse with for a given Neuron</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neighbor_reset">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neighbor_reset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neighbor_reset" title="Permalink to this definition">¶</a></dt>
<dd><p>This function deletes all the neighbor relationships in the connectome</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neighboring_blocks">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neighboring_blocks</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">block_ref</span></em>, <em class="sig-param"><span class="n">kernel_size</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neighboring_blocks" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the list of block ids who are neighbor of the given one
Block_id is in form of [x,y,z]</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neurons_in_block_neighborhood">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neurons_in_block_neighborhood</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">block_ref</span></em>, <em class="sig-param"><span class="n">kernel_size</span><span class="o">=</span><span class="default_value">3</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neurons_in_block_neighborhood" title="Permalink to this definition">¶</a></dt>
<dd><p>Provides the list of all neurons within the surrounding blocks given the kernel size with default being 3</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.neurons_in_the_block">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">neurons_in_the_block</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">block_ref</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.neurons_in_the_block" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a list of Neurons in the given block
block_id to be entered as [x,y,z]</p>
</dd></dl>

<dl class="py function">
<dt id="src.evo.synapse.synapse">
<code class="sig-prename descclassname">src.evo.synapse.</code><code class="sig-name descname">synapse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cortical_area</span></em>, <em class="sig-param"><span class="n">src_id</span></em>, <em class="sig-param"><span class="n">dst_cortical_area</span></em>, <em class="sig-param"><span class="n">dst_id</span></em>, <em class="sig-param"><span class="n">postsynaptic_current</span></em><span class="sig-paren">)</span><a class="headerlink" href="#src.evo.synapse.synapse" title="Permalink to this definition">¶</a></dt>
<dd><p>Function responsible for creating a synapse between a neuron and another one. In reality a single neuron can
have many synapses with another individual neuron. Here we use synaptic strength to simulate the same.
Note: Synapse association is captured on the Source Neuron side within Connectome</p>
<p># Input: The id for source and destination Neuron plus the parameter defining connection strength
# Source provides the Axon and connects to Destination Dendrite
# postsynaptic_current is intended to provide the level of synaptic strength</p>
</dd></dl>

</div>
<div class="section" id="module-src.evo">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-src.evo" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="src.inf.html" class="btn btn-neutral float-right" title="src.inf package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="src.edu.html" class="btn btn-neutral float-left" title="src.edu package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, Mohammad Nadji-Tehrani.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>